### PR TITLE
[quant][graphmode][fx] Support sigmoid/hardsigmoid/tanh in qat

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46738 [quant][graphmode][fx] Support sigmoid/hardsigmoid/tanh in qat**

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24486972](https://our.internmc.facebook.com/intern/diff/D24486972)

### CODE REVIEW

File: test/quantization/test_quantize_fx.py [hunk 1]
```diff
@@ -1458,10 +1499,17 @@ def forward(self, x):
 
         # This model is not executable since we just put all ops
         # in the same forward
-        m = M().eval()
+        m = M().train()
         # nothing to fuse so skipping the fuse step
-        qconfig_dict = {'': default_qconfig}
-        prepared = prepare_fx(m, qconfig_dict)
+        qconfig_dict = {'': default_qat_qconfig}
+        prepared = prepare_qat_fx(m, qconfig_dict)
```
Comment (raghuramank100) [inline test/quantization/test_quantize_fx.py:1505]: What happens for the case of PTQ? We should not be inserting observers in that case. Is that being tested?

File: torch/quantization/fx/pattern_utils.py [hunk 2]
```diff
@@ -14,17 +14,30 @@ def get_default_fusion_patterns():
...
+# e.g. pattern: torch.sigmoid,
+#      output_activation_post_process: default_affine_fixed_qparam_fake_quant
+DEFAULT_OUTPUT_ACTIVATION_POST_PROCESS_MAP = dict()
+
 # Register pattern for both static quantization and qat
-def register_quant_pattern(pattern):
+def register_quant_pattern(pattern, output_activation_post_process=None):
     def insert(fn):
         DEFAULT_QUANTIZATION_PATTERNS[pattern] = fn
+        if output_activation_post_process is not None:
```
Comment (raghuramank100) [inline torch/quantization/fx/pattern_utils.py:26]: When would output_activation_post_process be None?

File: test/quantization/test_quantize_fx.py (hunk 1 above)
Comment (jerryzh168) [reply test/quantization/test_quantize_fx.py:1505]: we'll insert fixedqparamfakequant in that case as well

File: torch/quantization/fx/quantize.py [hunk 3]
```diff
@@ -346,7 +345,7 @@ def _prepare(self, model, qconfig_dict, inplace, prepare_custom_config_dict, is_
 
         # find _inputs_ to matched nodes that are not quantized, these
         # have to be quantized, which requires measuring stats,
-        # initialize an DefaultQuant object for each
+        # initialize an DefaultQuantizeHandler object for each
```
Comment (raghuramank100) [inline torch/quantization/fx/quantize.py:348]: The function _prepare is really long, can you refactor it into smaller functions in a later PR?

File: test/quantization/test_quantize_fx.py (hunk 1 above)
Comment (jerryzh168) [reply test/quantization/test_quantize_fx.py:1505]: I can add a test for ptq

File: torch/quantization/fx/pattern_utils.py (hunk 2 above)
Comment (jerryzh168) [reply torch/quantization/fx/pattern_utils.py:26]: this is only used for specifying special activaiton_post_process (e.g. fixed qparam fake quant) for a pattern, it is None for all other cases.

File: test/quantization/test_quantize.py [hunk 4]
```diff
@@ -1247,6 +1248,36 @@ def forward(self, x):
...
+            def __init__(self):
+                super().__init__()
+                self.sigmoid = torch.nn.Sigmoid()
+                self.hardsigmoid = torch.nn.Hardsigmoid()
+                self.tanh = torch.nn.Tanh()
+                self.quant = QuantStub()
+                self.dequant = DeQuantStub()
+
+            def forward(self, x):
+                x = self.quant(x)
```
Comment (raghuramank100) [inline test/quantization/test_quantize.py:1264]: Does the approach here also work for cases where the scale and zero-point are not fixed? i.e for a hardTanh, the scale and zero-point depend on the arguments to init.

File: torch/quantization/quantize.py [hunk 5]
```diff
@@ -9,14 +9,23 @@
...
+    get_default_special_act_post_process,
+)
 
 from .stubs import DeQuantStub, QuantWrapper
 from .qconfig import default_dynamic_qconfig, float16_dynamic_qconfig, float_qparams_dynamic_qconfig
 
+def is_activation_post_process(module):
+    return (isinstance(module, torch.quantization.ObserverBase) or
+            isinstance(module, torch.quantization.FakeQuantize) or
+            isinstance(module, torch.quantization.FakeQuantizeBase))
```
Comment (supriyar) [inline torch/quantization/quantize.py:27]: Do we plan to make FakeQuantize a subclass of FakeQuantizeBase?

File: torch/quantization/fx/quantize.py (hunk 3 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:348]: sure

File: test/quantization/test_quantize.py (hunk 4 above)
Comment (jerryzh168) [reply test/quantization/test_quantize.py:1264]: maybe, this allows to provide special fake quantize(activation_post_process) for a specific pattern (module/functional/torch op), could you write down the details, is it like following?
```
fake_quant = HardTanhFakeQuantize(hardtanh_instance.params)
```
I think this should be too hard to support in current implementation.

File: torch/quantization/quantize.py (hunk 5 above)
Comment (jerryzh168) [reply torch/quantization/quantize.py:27]: yes, will do that sometime in the future, maybe before end of this year.

File: torch/quantization/quantize.py [hunk 6]
```diff
@@ -107,31 +116,34 @@ def add_observer_(module, qconfig_propagation_list=None, non_leaf_module_list=No
...
             # Register observer as the first entry in the hook list
             # All post forward hooks are preserved and will be executed after the observer before convert
             handle = register_activation_post_process_hook(m)
             m._forward_hooks.move_to_end(handle.id, last=False)
 
     for name, child in module.named_children():
         if type(child) == nnq.FloatFunctional or type(child) == nnq.QFunctional:
-            if hasattr(child, 'qconfig') and child.qconfig is not None:
+            if needs_observation(child):
                 child.activation_post_process = get_activation_post_process(child.qconfig, device)
```
Comment (supriyar) [inline torch/quantization/quantize.py:143]: I think this callsite needs to be updated.

File: test/quantization/test_quantize_fx.py [hunk 7]
```diff
@@ -1488,6 +1471,82 @@ def forward(self, x):
...
+                m.train()
+                qconfig = default_qat_qconfig
+                prepare = prepare_qat_fx
+
+            # nothing to fuse so skipping the fuse step
+            qconfig_dict = {'': qconfig}
+            prepared = prepare(m, qconfig_dict)
+            # check the correct number of activation_post_process is inserted
+            count_check = {
+                ns.call_module(FixedQParamsFakeQuantize) : 13,
```
Comment (supriyar) [inline test/quantization/test_quantize_fx.py:1522]: Do we want to use the `FixedQParamsFakeQuantize` for PTQ as well? FakeQuantize is something we typically use during QAT.

Comment (supriyar) [reply test/quantization/test_quantize_fx.py:1522]: Is there a way to fix scale and zero-point for PTQ without using this module? Or should we have a similar `FixedQParamsObserver`?

File: torch/quantization/quantize.py (hunk 6 above)
Comment (jerryzh168) [reply torch/quantization/quantize.py:143]: why? this is handling FloatFunctional and we don't have special activation post process for it.

Comment (jerryzh168) [reply torch/quantization/quantize.py:143]: oh, I guess I forgot about setting default argument for special_act_post_process for get_activation_post_process.

File: test/quantization/test_quantize_fx.py (hunk 7 above)
Comment (jerryzh168) [reply test/quantization/test_quantize_fx.py:1522]: yeah, actually, after second thought, I feel we don't need to insert anything for PTQ, since the information is not needed and also there is no point to have fake quantize module in the observed model.

Comment (jerryzh168) [reply test/quantization/test_quantize_fx.py]: Thanks Supriya, updated the PR.

Comment (supriyar) [reply test/quantization/test_quantize_fx.py]: Looks good. Could you clarify the flow for PTQ? How does the scale/zero_point for the fixed qparams ops get calculated?

File: test/quantization/test_quantize.py (hunk 4 above)
Comment (raghuramank100) [reply test/quantization/test_quantize.py:1264]: Actually for hardtanh, our implementation calls qclamp: https://github.com/pytorch/pytorch/blob/87a4baf616565aecd21884754772a0f03babfe2d/aten/src/ATen/native/quantized/cpu/qclamp.cpp#L24, which doesn't need LUTs. In this case, we should treat it like ReLU, i.e there is no fake-quant or observer needed at the output as there is no quantization that occurs inside the op.

**Review by raghuramank100 — APPROVED** [review]

File: test/quantization/test_quantize.py (hunk 4 above)
Comment (jerryzh168) [reply test/quantization/test_quantize.py:1264]: yeah this is how it is implemented right now: https://github.com/pytorch/pytorch/blob/master/torch/quantization/fx/quantization_patterns.py#L512
